src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -475,6 +475,45 @@ SHAMap::firstBelow(
     }
     return nullptr;
 }
+SHAMapLeafNode*
+SHAMap::lastBelow(
+    std::shared_ptr<SHAMapTreeNode> node,
+    SharedPtrNodeStack& stack,
+    int branch) const
+{
+    // Return the first item at or below this node
+    if (node->isLeaf())
+    {
+        auto n = std::static_pointer_cast<SHAMapLeafNode>(node);
+        stack.push({node, {leafDepth, n->peekItem()->key()}});
+        return n.get();
+    }
+    auto inner = std::static_pointer_cast<SHAMapInnerNode>(node);
+    if (stack.empty())
+        stack.push({inner, SHAMapNodeID{}});
+    else
+        stack.push({inner, stack.top().second.getChildNodeID(branch)});
+    for (int i = branchFactor; i >= 0;)
+    {
+        if (!inner->isEmptyBranch(i))
+        {
+            node = descendThrow(inner, i);
+            assert(!stack.empty());
+            if (node->isLeaf())
+            {
+                auto n = std::static_pointer_cast<SHAMapLeafNode>(node);
+                stack.push({n, {leafDepth, n->peekItem()->key()}});
+                return n.get();
+            }
+            inner = std::static_pointer_cast<SHAMapInnerNode>(node);
+            stack.push({inner, stack.top().second.getChildNodeID(branch)});
+            i = branchFactor;  // scan all 16 branches of this new node
+        }
+        else
+            --i;  // scan next branch
+    }
+    return nullptr;
+}
```

Comment:
This new `lastBelow` function is almost identical to `firstBelow`. We could use a single implementation. Something like (untested, but you get the idea):
```
namespace detail {
struct Forward
{
    static int
    start()
    {
        return 0;
    }
    static int
    end()
    {
        return SHAMap::branchFactor;
    }
    static int
    next(int i)
    {
        return i + 1;
    }
};
struct Reverse
{
    static int
    start()
    {
        return SHAMap::branchFactor - 1;
    }
    static int
    end()
    {
        return -1;
    }
    static int
    next(int i)
    {
        return i - 1;
    }
};
}  // namespace detail

template <class I>
SHAMapLeafNode*
SHAMap::belowHelper(
    std::shared_ptr<SHAMapTreeNode> node,
    SharedPtrNodeStack& stack,
    int branch) const
{
    // Return the first item at or below this node
    if (node->isLeaf())
    {
        auto n = std::static_pointer_cast<SHAMapLeafNode>(node);
        stack.push({node, {leafDepth, n->peekItem()->key()}});
        return n.get();
    }
    auto inner = std::static_pointer_cast<SHAMapInnerNode>(node);
    if (stack.empty())
        stack.push({inner, SHAMapNodeID{}});
    else
        stack.push({inner, stack.top().second.getChildNodeID(branch)});
    for (int i = I::start(); i != I::end();)
    {
        if (!inner->isEmptyBranch(i))
        {
            node = descendThrow(inner, i);
            assert(!stack.empty());
            if (node->isLeaf())
            {
                auto n = std::static_pointer_cast<SHAMapLeafNode>(node);
                stack.push({n, {leafDepth, n->peekItem()->key()}});
                return n.get();
            }
            inner = std::static_pointer_cast<SHAMapInnerNode>(node);
            stack.push({inner, stack.top().second.getChildNodeID(branch)});
            i = I::start();  // scan all 16 branches of this new node
        }
        else
            i = I::next(i);  // scan next branch
    }
    return nullptr;
}

SHAMapLeafNode*
SHAMap::firstBelow(
    std::shared_ptr<SHAMapTreeNode> node,
    SharedPtrNodeStack& stack,
    int branch) const
{
    return belowHelper<detail::Forward>(std::move(node), stack, branch);
}
SHAMapLeafNode*
SHAMap::lastBelow(
    std::shared_ptr<SHAMapTreeNode> node,
    SharedPtrNodeStack& stack,
    int branch) const
{
    return belowHelper<detail::Reverse>(std::move(node), stack, branch);
}

```

BTW, I don't object to two functions either. Sometimes more generic is harder to understand and debug. If you prefer using two function I'm fine with that. The above is just a thought.